### PR TITLE
Implement iOS authentication and streams

### PR DESCRIPTION
**Pull request by fotiDim**
This PR implements the iOS authentication functions and establishes subscriptions to streams.
Check the README to see what exactly is included.
Some open points still exist and need to be thought of. Please see the inline comments in the PR.

### CODE REVIEW

File: ios/Classes/SwiftSpotifySdkPlugin.swift [hunk 1]
```diff
@@ -3,63 +3,107 @@ import SpotifyiOS
...
+                }
+                guard let playerState = playerState as? SPTAppRemotePlayerState else {
+                    result(FlutterError(code: "PlayerAPI Error", message: "PlayerState is empty", details: nil))
+                    return
+                }
+                result(PlayerState.stateJson(playerState).json)
+            })
+        case SpotfySdkConstants.methodLogoutFromSpotify:
+            appRemote?.disconnect()
+//            appRemote?.connectionParameters.accessToken = nil
```
Comment (fotiDim) [inline ios/Classes/SwiftSpotifySdkPlugin.swift:85]: This is the most controversial point of the PR. According to my research there is no logout for the Android or iOS SDK. For now I am treating it as `disconnect()` and from what I understand so does the Android implementation.

I suggest either to:
- Rename the method to disconnect()
- Keep that and create an additional disconnect() method. The difference would be that on iOS `LogoutFromSpotify()` would also clear the stored `accessToken`. On Android from what I saw the token is not persisted at all.

Thoughts?

File: lib/spotify_sdk.dart [hunk 2]
```diff
@@ -46,18 +46,21 @@ class SpotifySdk {
...
-      String playerName = 'Spotify SDK'}) async {
+      String playerName = 'Spotify SDK',
+      String accessToken}) async {
     try {
       return await _channel.invokeMethod(MethodNames.connectToSpotify, {
         ParamNames.clientId: clientId,
         ParamNames.redirectUrl: redirectUrl,
-        ParamNames.playerName: playerName
+        ParamNames.playerName: playerName,
+        ParamNames.accessToken: accessToken
```
Comment (fotiDim) [inline lib/spotify_sdk.dart:63]: On iOS if you pass a valid accessToken during connection it avoids having to switch to the Spotify app. Also documented in the README.

File: ios/Classes/SwiftSpotifySdkPlugin.swift (hunk 1 above)
Comment (itsMatoosh) [reply ios/Classes/SwiftSpotifySdkPlugin.swift:93]: The logout function doesn't exist in the Spotify Web SDK either. When writing the Web implementation of this plugin, I also made the choice to treat the logout() function as a disconnect. The logout() method should definitely be renamed to disconnect().

Comment (fotiDim) [reply ios/Classes/SwiftSpotifySdkPlugin.swift:93]: @itsMatoosh thanks for the input. I was also wondering.

Comment (brim-borium) [reply ios/Classes/SwiftSpotifySdkPlugin.swift:93]: @fotiDim do you want do rename this here or in another PR ?

Comment (fotiDim) [reply ios/Classes/SwiftSpotifySdkPlugin.swift:93]: Another pr would be easier

Comment (brim-borium) [reply ios/Classes/SwiftSpotifySdkPlugin.swift:93]: Alright Merging this then 👍

**Review by brim-borium — APPROVED** [review]

**Comment by itsMatoosh** [comment]
I'm having issues with the authorization methods which require opening the Spotify app (connectToSpotifyRemote and getAuthenticationToken). When called, the Spotify application opens, shows the "Authorizing" popup, and goes back to my app. The methods, however, never resolve, and no result is ever presented. I'm not sure if I set something up wrong, or if this is a bug.

**Comment by fotiDim** [comment]
@itsMatoosh does the demo app work for you?
With your app, when the Spotify app completes "Authorizing" does it show an 𝗫 or a ✔️ before redirecting back to your app?

I saw a path in the code where if you don't get a token back from Spotify then there is no error reported back to the Flutter side. It should not block you though from being able to connect successfully.

I made the connection code in my last PR a bit more defensive. Can you check out the corresponding branch and see if you are getting any meaningful error now?

**Comment by itsMatoosh** [comment]
I tried using your last PR, and I'm still having the same issue. Spotify app opens, shows "Authorizing", then ✔️, then my app opens, but no result is returned. It seems like the ios part responsible for handling the response url is never called.

**Comment by fotiDim** [comment]
Can you run the project through Xcode and put a breakpoint in line 249 in SwiftSpotifySdkPlugin.swift?
Is `setAccessTokenFromURL()` even called?

I am also available in the Slack channel.

**Comment by itsMatoosh** [comment]
I put a breakpoint at line 249 (result?(token)) and it never gets called. How do I contact you on slack?

**Comment by fotiDim** [comment]
I see you have an invite already for `spotifysdk.slack.com`. You need to accept it.

**Pull request merged**